Mathlib/RingTheory/TensorProduct.lean

```diff
@@ -590,19 +611,30 @@ theorem lid_tmul (r : R) (a : A) : (TensorProduct.lid R A : R ⊗ A → A) (r 
   simp [TensorProduct.lid]
 #align algebra.tensor_product.lid_tmul Algebra.TensorProduct.lid_tmul
 
+variable (S)
+
 /-- The base ring is a right identity for the tensor product of algebra, up to algebra isomorphism.
+
+Note that if `A` is commutative this can be instantiated with `S = A`.
 -/
-protected nonrec def rid : A ⊗[R] R ≃ₐ[R] A :=
-  algEquivOfLinearEquivTensorProduct (TensorProduct.rid R A) (by
-    simp [mul_smul]
-    simp_rw [← mul_smul, mul_comm]
-    simp)
-    (by simp [Algebra.smul_def])
+protected nonrec def rid : A ⊗[R] R ≃ₐ[S] A :=
+  let act : R →ₗ[R] A →ₗ[S] A :=
+    (Algebra.lsmul (A := A) S Aᵐᵒᵖ A).toLinearMap.flip.restrictScalars R ∘ₗ Algebra.linearMap R A
+  let lin : A ⊗[R] R ≃ₗ[S] A := LinearEquiv.ofLinear
+    (AlgebraTensorModule.lift <| LinearMap.flip <| act)
+    ((AlgebraTensorModule.mk R A A R).flip 1)
+    (LinearMap.ext <| fun x => show x * algebraMap R A 1 = x by simp)
+    (AlgebraTensorModule.ext <| fun x y => show (x * algebraMap R A y) ⊗ₜ[R] 1 = x ⊗ₜ[R] y
+      by rw [←Algebra.commutes, ←_root_.Algebra.smul_def, smul_tmul, smul_eq_mul, mul_one])
+  algEquivOfLinearEquivTensorProduct lin
+    (fun a₁ a₂ r₁ r₂ => by
+      show a₁ * a₂ * algebraMap R A (r₁ * r₂) = a₁ * algebraMap R A r₁ * (a₂ * algebraMap R A r₂)
+      simp_rw [←Algebra.commutes, ←Algebra.smul_def, smul_mul_smul])
+    (fun s => show algebraMap S A s * algebraMap R A 1 = algebraMap S A s by simp)
 #align algebra.tensor_product.rid Algebra.TensorProduct.rid
 
 @[simp]
-theorem rid_tmul (r : R) (a : A) : (TensorProduct.rid R A : A ⊗ R → A) (a ⊗ₜ r) = r • a := by
-  simp [TensorProduct.rid]
+theorem rid_tmul (r : R) (a : A) : TensorProduct.rid R S A (a ⊗ₜ r) = a * algebraMap _ _ r := rfl
```

Comment:
Is this change reasonable? Or was the old defeq better?

Reply:
I've pinged Kevin and Antoine on Zulip just to be safe; I'll probably merge tomorrow if they don't reply before then.

Reply:
I have a slight preference for the old version, but I can't bet it is reasonable.
According to my own programming habit, keeping`r • a` (smul) as long as possible is safer, and one can always rewrite it to multiplication using `algebraMap_eq_smul`.

Reply:
@AntoineChambert-Loir, note this is `a * algebraMap _ _ r` not `algebraMap _ _ r * a`; though admittedly that's only one more rewrite away.

Reply:
I've changed it to `r • a`